src/cart/crt_context.c

```diff
@@ -1202,50 +1362,49 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
 	}
 
 	/* process waitq */
-	inflight = epi->epi_req_num - epi->epi_reply_num;
-	D_ASSERT(inflight >= 0 && inflight <= crt_gdata.cg_credit_ep_ctx);
-	credits = crt_gdata.cg_credit_ep_ctx - inflight;
-	while (credits > 0 && !d_list_empty(&epi->epi_req_waitq)) {
+	while (credits_available(epi) > 0 && !d_list_empty(&epi->epi_req_waitq)) {
 		D_ASSERT(epi->epi_req_wait_num > 0);
-		tmp_rpc = d_list_entry(epi->epi_req_waitq.next,
-					struct crt_rpc_priv, crp_epi_link);
-		tmp_rpc->crp_state = RPC_STATE_INITED;
-		crt_set_timeout(tmp_rpc);
+		tmp_rpc = d_list_entry(epi->epi_req_waitq.next, struct crt_rpc_priv, crp_epi_link);
+		RPC_ADDREF(tmp_rpc);
+		D_MUTEX_UNLOCK(&epi->epi_mutex);
 
-		D_MUTEX_LOCK(&crt_ctx->cc_mutex);
-		rc = crt_req_timeout_track(tmp_rpc);
-		D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
-		if (rc != 0)
-			RPC_ERROR(tmp_rpc,
-				"crt_req_timeout_track failed, rc: %d.\n", rc);
+		D_MUTEX_LOCK(&tmp_rpc->crp_mutex);
+		D_MUTEX_LOCK(&epi->epi_mutex);
+		if (tmp_rpc->crp_state == RPC_STATE_QUEUED && credits_available(epi) > 0) {
+			tmp_rpc->crp_state = RPC_STATE_INITED;
 
```

Comment:
original calling of "crt_req_timeout_track" removed, not quite sure you mean it is unnecessary?

Reply:
Yes, this patch changes crt_context_req_track to call crt_req_timeout_track for both waiting and inflight requests. Hence, the waiting request here should already be in the timeout binheap.

Reply:
ok, original code will not add to timeout binheap if it is queued, but seems not bad to track timeout for QUEUED case also.

Reply:
It does have some downside: I guess we didn't track timeouts for waiting requests because we wanted to avoid timeouts due to local waits. I'm also a little concerned if we will see such timeouts as a result of this patch.

On the other hand, tracking timeouts for waiting requests makes it possible for users to limit the time spent on waiting for an RPC. For example, in the future, if we track the deadline of an incoming RPC in an engine, and the RPC handler needs to send an other RPC to a different engine, then we may want to get a -DER_TIMEDOUT for the outgoing RPC before the deadline of the incoming RPC.

Personally, I'm not sure which way is _definitely_ better, and would be happy to revert this change if you guys prefer the current behavior.

Reply:
@frostedcmos also commented on this change. Since everyone, including myself, turns out to be a bit concerned, I'm working on restoring the previous behavior.

Reply:
Restored to the previous behavior (i.e., not starting timer until inflight).